src/Language/PureScript/Sugar/BindingGroups.hs

```diff
@@ -76,18 +76,18 @@ createBindingGroups moduleName = mapM f <=< handleDecls
         kindDecls = fmap (,VertexKindSignature) $ filter (\a -> isKindDecl a || isExternDataDecl a) ds
         dataDecls = fmap (,VertexDefinition) $ filter (\a -> isDataDecl a || isTypeSynonymDecl a || isTypeClassDecl a) ds
         kindSigs = fmap (declTypeName . fst) kindDecls
-        typeSyns = fmap declTypeName $ filter isTypeSynonymDecl ds
+        kindedData = (fmap declTypeName $ filter (\a -> isDataDecl a || isExternDataDecl a) ds) `intersect` kindSigs
```

Comment:
This change may be an oversimplification. I don't think data types or instances need to depend on a class definition. I think this may affect recursive binding groups that contains classes and polykinds. I don't know if we have a test for that or not.

Reply:
I think it would be worth noting why super classes depend on the class definition, and if it's truly necessary.

Reply:
Honestly, I don't understand why this can't be *further* simplified to always depend on the definition for everything referenced (which will itself continue to depend on its own kind declaration, if present). I assumed there was a reason why depending on a data type's kind declaration instead of the data type itself would be necessary, but we don't seem to have any tests exercising that—nothing fails when I change that logic, aside from a trivial unknown renumbering in one of the golden outputs.

For recursive binding groups containing classes and polykinds, we have [tests/purs/passing/PolykindBindingGroup2.purs](https://github.com/purescript/purescript/blob/bc898395c2da556c8ec6b7c702a3dff6d2f1bb9c/tests/purs/passing/PolykindBindingGroup2.purs); is that the sort of thing you had in mind, but maybe with some kind declarations thrown in?

At some level, none of this reordering is ‘truly necessary’; the type checker itself could always have logic to scan all of the declarations or do its own topsorting or whatever, right? Shouldn't the burden of proof instead be on the desugaring to generally put things in order of reference or in recursive binding groups and deviate from that only when truly necessary?

Reply:
It's required to support polymorphic recursion (which requires polymorphic kind signatures). The type checker for value declarations essentially does the same thing by splitting binding groups into things which have explicit signatures and things that don't. We don't do that for top-level kind signatures because there's no desugaring to remove the top-level signature like there is for values and their type declarations. It may be possible to have a different architecture for handling explicit kind signatures in these cases, but you will have to make fairly significant changes to the environment, I think.

Reply:
I fully believe you but I'd like to have an example to add to the tests that fails if the compiler does the naive thing I proposed. I've tried a few cases of polymorphic recursion but none of them have fit the bill yet; could I trouble you to suggest a specific example?

Reply:
> It's required to support polymorphic recursion (which requires polymorphic kind signatures). 

This is not correct, and I'm conflating two different things from what I remember from the original paper. For reference https://dl.acm.org/doi/pdf/10.1145/3371121 6.1 is the brief mention of this. For the case of polymorphic recursion, it would be enough for kind signatures to be in scope at the point checking the other declarations, which would be equivalent to what the value-level type checker is doing. The only thing that is important here is to catch and disallow cyclical dependencies in kind signatures.

I _think_ that as long as:
* Kind signatures are always in scope at the point of checking the declaration (ie, the declaration depends on it's kind signature).
* Cycles in kind signatures are disallowed.

Then alternative orderings should be acceptable. Depending on only the kind signature if available does conceptually simplify things, since it will break up recursive groups.

Reply:
That's very helpful, thanks!

So yeah, it seems like it's not important to break up groups containing definitions—these groups would have existed before kind signatures, and the type checker must have handled them—but it is very important to make sure that kind signatures don't get involved in those groups. So the algorithm I'm pushing now has kind signatures depend on other kind signatures if present (except for type synonyms) and everything else depend on definitions. I see where Xie et al. recommend depending on kind signatures more broadly, but their type system doesn't include type classes, and they don't seem to make a strong argument for the benefit of that approach aside from keeping kind signatures out of groups. I'm adding a comment describing this as a potential future approach if needed and if the type checker can be made to cooperate.